src/components/Form.js

```diff
@@ -139,11 +139,9 @@ export default class Form extends Component {
         errors: toErrorList(newErrorSchema),
       };
     }
-    setState(this, state, () => {
-      if (this.props.onChange) {
-        this.props.onChange(this.state);
-      }
-    });
+
+    this.setState(state);
```

Comment:
I'm wondering, why not do something like this:

```suggestion
    this.setState(state, () => this.props.onChange && this.props.onChange(state) );
```

Reply:
Oh so that's what you and @sbusch meant by race condition -- using the callback to call props.onChange after a re-render completes/update to state has been applied. As noted by the docs (https://reactjs.org/docs/react-component.html#setstate) the pitfall is when you're `this.state` since it won't reflect the change you're about to/attempting to do. However, as this an instanced variable not tied to `this.state`, I don't think such a "pitfall" is possible here. Instead, what I think would happen is that the `setState` may defer the update to the component until the 1st thread's execution is complete and, thus, `props.onChange` may be called before the `Form` updates its own state. Consequentially the parent/user could end up acting on a change event and change the props of the `Form` initiating another update to the form (based on the future state) -- so the race condition could be that state updates with a `newProps` with `this.state = oldState`, and then updates again with just the new state (potentially discarding any derived state updates from a change in props and the new current state (which was pretty much the bug I had mentioned when I used `props.onChange` before `this.setState()`).

I think not using the callback and allowing React to *batch* the updates (to the just the state I believe -- it can't possible have a reference to oldProps and newState, right? https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous) is fine (though I'm no React expert, just giving my 2 cents) if there aren't such hooks that compare `newProps` vs `oldProps` and the current `this.state` (though that's what one of my Form wrappers has to do and caused an issue where it reset the values).

So sure, I guess using callback to slow it down and force everything to be synchronous may be best/least messy course of action, even if it potentially increases the number of times render() is called.